src/main/scala/is/hail/rvd/OrderedRVDPartitioner.scala

```diff
@@ -8,66 +8,144 @@ import org.apache.spark.Partitioner
-
-    val part = BinarySearch.binarySearch(numPartitions,
-      // key.compare(elem)
-      i =>
-        if (i == numPartitions - 1)
-          -1 // key.compare(inf)
-        else
-          ordering.compare(pk, rangeBounds(i)))
-    part
+    val part = rangeTree.queryValues(pkType.ordering, pk)
+    part.length match {
+      case 0 =>
+        if (pkType.ordering.gt(pk, rangeTree.root.get.maximum))
+          numPartitions - 1
+        else {
+          assert(pkType.ordering.lt(pk, rangeTree.root.get.minimum))
+          0
+        }
+      case 1 => part.head
+    }
   }
 
-  // return the smallest partition for which key <= max
-  // key: RegionValue[kType]
+  // return the partition containing pk
+  // needs to update the bounds
+  // key: RegionValue
   def getPartition(key: Any): Int = {
     val keyrv = key.asInstanceOf[RegionValue]
+    val keyUR = new UnsafeRow(kType, keyrv)
+
+    val part = rangeTree.queryValues(pkKOrd, keyUR)
 
-    val part = BinarySearch.binarySearch(numPartitions,
-      // key.compare(elem)
-      i =>
-        if (i == numPartitions - 1)
-          -1 // key.compare(inf)
-        else
-          -pkKOrd.compare(rangeBounds.region, loadElement(i), keyrv))
-    part
+    part.length match {
+      case 0 =>
+        if (pkKOrd.lt(rangeTree.root.get.maximum, keyUR))
+          numPartitions - 1
+        else {
+          assert(pkKOrd.gt(rangeTree.root.get.minimum, keyUR))
+          0
+        }
+      case 1 => part.head
+    }
   }
 
   def withKType(newPartitionKey: Array[String], newKType: TStruct): OrderedRVDPartitioner = {
     val newPart = new OrderedRVDPartitioner(numPartitions, newPartitionKey, newKType, rangeBounds)
     assert(newPart.pkType == pkType)
     newPart
   }
+
+  def copy(numPartitions: Int = numPartitions, partitionKey: Array[String] = partitionKey,
+    kType: TStruct = kType, rangeBounds: UnsafeIndexedSeq = rangeBounds): OrderedRVDPartitioner = {
+    new OrderedRVDPartitioner(numPartitions, partitionKey, kType, rangeBounds)
+  }
+
+  def remapRanges(newPartEnd: Array[Int]): OrderedRVDPartitioner = {
+    val newRangeBounds = UnsafeIndexedSeq(
+      TArray(TInterval(pkType)),
+      (-1 +: newPartEnd.init).zip(newPartEnd).map { case (s, e) =>
+        val i1 = rangeBounds(s + 1).asInstanceOf[Interval]
+        val i2 = rangeBounds(e).asInstanceOf[Interval]
+        Interval(i1.start, i2.end, i1.includeStart, i2.includeEnd)
+      })
+    copy(numPartitions = newPartEnd.length, rangeBounds = newRangeBounds)
+  }
 }
 
 object OrderedRVDPartitioner {
   def empty(typ: OrderedRVDType): OrderedRVDPartitioner = {
     new OrderedRVDPartitioner(0, typ.partitionKey, typ.kType, UnsafeIndexedSeq.empty(TArray(typ.pkType)))
   }
+
+  // takes npartitions + 1 points and returns npartitions intervals: [a,b], (b,c], (c,d], ... (i, j]
+  def makeRangeBoundIntervals(pType: Type, rangeBounds: Array[RegionValue]): UnsafeIndexedSeq = {
+    val newT = TArray(TInterval(pType))
```

Comment:
Just call `makeRangeBoundIntervals(UnsafeIndexedSeq(TArray(pType), rangeBounds))`?  The duplicate pains me.